pinot-core/src/main/java/org/apache/pinot/core/query/executor/TableExecutionInfo.java

```diff
@@ -0,0 +1,197 @@
+   */
+  int getNumSegmentsAcquired();
+
+  /**
+   * Get the selected segments and segment contexts for a table referenced in a query. The information is gathered
+   * in a SelectSegmentsInfo object.
+   * @param queryContext
+   * @param timerContext
+   * @param executorService
+   * @param segmentPrunerService
+   * @return A SelectSegmentsInfo object containing the selected segments and segment contexts
+   */
+  SelectedSegmentsInfo getSelectedSegmentsInfo(QueryContext queryContext, TimerContext timerContext,
+      ExecutorService executorService, SegmentPrunerService segmentPrunerService);
+
+  /**
+   * Generate metadata about the consuming segments queried.
+   * @return A ConsumingSegmentsInfo object containing the metadata about the consuming segments queried
+   */
+  ConsumingSegmentsInfo getConsumingSegmentsInfo();
+
+  /**
+   * If consuming segments are being queried, this class contains the information about the consuming segments such as
+   * the number of segments queried, the min index time, the min ingestion time and the max end time.
+   */
+  class ConsumingSegmentsInfo {
+    private final int _numConsumingSegmentsQueried;
+    private final long _minIndexTimeMs;
+    private final long _minIngestionTimeMs;
+    private final long _maxEndTimeMs;
+
+    public ConsumingSegmentsInfo(int numConsumingSegmentsQueried, long minIndexTimeMs, long minIngestionTimeMs,
+        long maxEndTimeMs) {
+      _numConsumingSegmentsQueried = numConsumingSegmentsQueried;
+      _minIndexTimeMs = minIndexTimeMs;
+      _minIngestionTimeMs = minIngestionTimeMs;
+      _maxEndTimeMs = maxEndTimeMs;
+    }
+
+    long getMinConsumingFreshnessTimeMs() {
+      long minConsumingFreshnessTimeMs = 0;
+      if (getMinIngestionTimeMs() != Long.MAX_VALUE) {
+        minConsumingFreshnessTimeMs = getMinIngestionTimeMs();
+      } else if (getMinIndexTimeMs() != Long.MAX_VALUE) {
+        minConsumingFreshnessTimeMs = getMinIndexTimeMs();
+      } else if (getMaxEndTimeMs() != Long.MIN_VALUE) {
+        minConsumingFreshnessTimeMs = getMaxEndTimeMs();
+      }
+      return minConsumingFreshnessTimeMs;
+    }
+
+    public int getNumConsumingSegmentsQueried() {
+      return _numConsumingSegmentsQueried;
+    }
+
+    public long getMinIndexTimeMs() {
+      return _minIndexTimeMs;
+    }
+
+    public long getMinIngestionTimeMs() {
+      return _minIngestionTimeMs;
+    }
+
+    public long getMaxEndTimeMs() {
+      return _maxEndTimeMs;
+    }
+  }
+
+  /**
+   * This class contains the information about the selected segments such as the number of segments queried, the
+   * number of segments selected and the number of total documents.
+   */
+  class SelectedSegmentsInfo {
+    public List<IndexSegment> _indexSegments;
+    public long _numTotalDocs;
+    public SegmentPrunerStatistics _prunerStats;
+    public int _numTotalSegments;
+    public int _numSelectedSegments;
+    public List<SegmentContext> _selectedSegmentContexts;
```

Comment:
Are these intentionally publicly accessible and mutable? I don't see where we're mutating these? Why not make them private members accessible through getter methods like in `ConsumingSegmentsInfo`?

Reply:
Done